### ISSUE TITLE
@ (at) operator not accepting calculated timestamp.

### ISSUE THREAD

**Issue by aamkye** [prometheus/prometheus#12318]
### What did you do?

I was trying to retrieve specific metrics at yesterdays midnight (23:59:59) by using something like this:
```
some_custom_metric @ <timestamp>
```
while 
```
(time() - (hour(vector(time())) * 60 * 60 + minute(vector(time())) * 60) - 60)
and
0+(time() - (hour(vector(time())) * 60 * 60 + minute(vector(time())) * 60) - 60)
```
returns an OK timestamp, but while trying to use it like

```
some_custom_metric @ (time() - (hour(vector(time())) * 60 * 60 + minute(vector(time())) * 60) - 60)
or
some_custom_metric @ 0+(time() - (hour(vector(time())) * 60 * 60 + minute(vector(time())) * 60) - 60)
```
I'm receiving these kinds of errors

```
Error executing query: 1:10: parse error: unexpected "(" in @, expected timestamp
or
Error executing query: 1:1: parse error: @ modifier must be preceded by an instant vector selector or range vector selector or a subquery
```

### What did you expect to see?

Metric at a specific time.

### What did you see instead? Under which circumstances?

Errors:
```
Error executing query: 1:10: parse error: unexpected "(" in @, expected timestamp
or
Error executing query: 1:1: parse error: @ modifier must be preceded by an instant vector selector or range vector selector or a subquery
```

### System information

Prometheus in K8s via Helm (2.43.0)

### Prometheus version

```text
Version	2.43.0
Revision	edfc3bcd025dd6fe296c167a14a216cab1e552ee
Branch	HEAD
BuildUser	root@8a0ee342e522
BuildDate	20230321-12:56:07
GoVersion	go1.19.7
```

### Prometheus configuration file

```yaml
-
```

### Alertmanager version

```text
-
```

### Alertmanager configuration file

```yaml
-
```

### Logs

```text
-
```

**Comment by beorn7** [comment]
There have been thoughts around this for a long time, see [this exploratory document](https://docs.google.com/document/d/1jMeDsLvDfO92Qnry_JLAXalvMRzMSB1sBr9V7LolpYM/edit#), some parts of which are already implemented, some are [being worked on](https://github.com/prometheus/prometheus/pull/9138), and some are somewhat out in the distance, among them arithmetic on durations and on the argument of the `@` modifier.

**Comment by NeerajGartia21** [comment]
Is it possible to currently work on this issue, or is it still somewhat in the distance? If possible I can have a look into it.

**Comment by beorn7** [comment]
I would say we shouldn't look at arithmetic for the `@` modifier in isolation, but for all durations.

The challenge here is that we need in a way a new concept of "static" arithmetic, i.e. evaluations that do not touch the TSDB. Otherwise, it would be possible to create expressions that require multiple subsequent TSDB lookups (which is something that we might want to support in the future, but not in this first step).

So for example, it should be possible to write
`rate(foo_total[1m + 2m] offset (1d + 1h))` but not `rate(foo_total[interval_set_in_recording_rule])`

This might be "not as easy as it looks", but if you would like to give it a spin, it would certainly be very welcome.

**Comment by NeerajGartia21** [comment]
> I would say we shouldn't look at arithmetic for the @ modifier in isolation, but for all durations.

So to achieve this we need to replace all the `Duration` component with `DurationArithmetic` component right? And also I think It will be better and easy to start with this one after #13412 gets merged.

> This might be "not as easy as it looks", but if you would like to give it a spin, it would certainly be very welcome.

Yeah, it looks complex but I'll try to look into it in my free time. I need to deep dive into the promql engine for this.

> It will touch the parser, for sure, but it might also require quite deep changes in the PromQL engine 

Yeah I think it will touch both parser as well as the engine and needs a deep understanding of promql engine I guess.

**Comment by beorn7** [comment]
Hello from the bug scrub!

This is still a valid issue. We changed the title to clarify that this is more than just about the @ argument.

**Comment by ringerc** [comment]
It looks like this was closed by https://github.com/prometheus/prometheus/pull/16249

**Comment by beorn7** [comment]
We do not have arithmetic on the parameter of the `@` modifier. (It's a timestamp, and not a duration, but I think it would still be interesting, as the original poster explained.)

@roidelapluie maybe we discussed that already, but would you think that arithmetic on the parameter of `@` would also be useful?

Also, I still get squiggly lines for duration expressions:
<img width="207" height="91" alt="Image" src="https://github.com/user-attachments/assets/716d7a0e-59fe-4249-8010-8107b69dc8fc" />

@juliusv @Nexucis is that expected? I have vague memory that something was done for frontend (autocompletion and squiggly lines), but maybe that's still to do?

**Comment by ringerc** [comment]
@beorn7 I strongly agree that arithmetic (and expression expansion with `step()` etc) on `@` offset modifiers would also be valuable, yes.

So this feature request is now partially addressed - for `[duration]`s but not for `@ offset`s

**Comment by juliusv** [comment]
> @juliusv @Nexucis is that expected? I have vague memory that something was done for frontend (autocompletion and squiggly lines), but maybe that's still to do?

I've always left these highlighting extensions to @Nexucis because he's the only one who properly understands the code around that 🙈

**Comment by Nexucis** [comment]
> @juliusv @Nexucis is that expected? I have vague memory that something was done for frontend (autocompletion and squiggly lines), but maybe that's still to do?

I think it was a TODO from a PR that added this syntax but the frontend was not modified to follow these changes

**Comment by Nexucis** [comment]
and it doesn't seem it will be a quick fix ... Including arithmetic in the duration is making the grammar ambiguous.

`1 offset 1d + 3` can be interpreted as `Expr Offset Duration` or `Expr offset Duration BinOp(Add) Expr` and lezer does not like that at all.

I am not even sure to understand how this issue has been solved in yacc
Any clue @roidelapluie or @beorn7 ?

**Comment by juliusv** [comment]
@Nexucis Looking at https://github.com/prometheus/prometheus/blob/main/promql/parser/generated_parser.y#L1038-L1244, I think this ambiguity is resolved by defining explicit separate patterns for the offset duration expressions (`offset_duration_expr`) here: https://github.com/prometheus/prometheus/blob/main/promql/parser/generated_parser.y#L1038-L1244

For example, after taking care of a few special cases for `offset_duration_expr` for unary operators, `step()`, `min()`, `max()`, etc., the grammar rule also allows a general `duration_expr` for the `offset_duration_expr`, and that rule defines its own binary operations like `ADD`, `SUB`, etc. specifically for offset expressions. And then I think the left associativity of the binary operators (see https://github.com/prometheus/prometheus/blob/25aee26a57ed08c8c4f518c56c834f6d2f567519/promql/parser/generated_parser.y#L194-L199) will cause `foo offset 1m + 1` to be interpreted as `(foo offset 1m) + 1)` instead of `foo offset (1m + 1)`, unless you explicitly set the parentheses like that.

**Comment by ringerc** [comment]
Since this is experimental, would it make sense to require that offset expressions with arithmetic always be parenthesised?

I know that's what PostgreSQL sometimes does in cases where ambiguous grammar is an issue. And it'd IMO be a good habit to encourage anyway.

**Comment by Nexucis** [comment]
ah cool thank you for the insight @juliusv.
I didn't think it required that much changes :/. 

@juliusv what do you think about simply removing the duration as a token (in the frontend) and we stop trying to autocomplete the keyword of the duration `d,m,s,ms` ?
Highlightning will still work but we don't care anymore if it's a number or a duration, everything is a number and that's it

**Comment by roidelapluie** [comment]
Well, offset_duration_expr was only needed for `^`, for `+` it worked directly.

**Comment by juliusv** [comment]
> ah cool thank you for the insight [@juliusv](https://github.com/juliusv). I didn't think it required that much changes :/.
> 
> [@juliusv](https://github.com/juliusv) what do you think about simply removing the duration as a token (in the frontend) and we stop trying to autocomplete the keyword of the duration `d,m,s,ms` ? Highlightning will still work but we don't care anymore if it's a number or a duration, everything is a number and that's it

It would be a little bit of a regression IMO, but if it gets generally impossible to tell what is supposed to be completed as a regular duration or as just a number (like you don't know if the user is going to type `5m * 5` or `5 * 5m`), maybe we just have to do that.
